delfin/tests/unit/drivers/huawei/oceanstor/test_oceanstor.py

```diff
@@ -328,3 +328,465 @@ def test_list_volumes(self):
+                    'TYPE': '214',
+                    'NAME': 'TEST_SAS_PORT',
+                    'RUNNINGSTATUS': '10',
+                    'HEALTHSTATUS': '1',
+                    'ID': '55555',
+                    'LOCATION': 'Location6',
+                    'RUNSPEED': '12000',
+                    'MAXSPEED': '12000',
+                    'PARENTID': '0A',
+                }],
+                'error': {
+                    'code': 0,
+                    'description': '0'
+                }
+            },
+            {
+                'data': [{
+                    'TYPE': '210',
+                    'ID': '012345',
+                    'NAME': 'Name100',
+                    'RUNNINGSTATUS': '27',
+                    'HEALTHSTATUS': '1',
+                }],
+                'error': {
+                    'code': 0,
+                    'description': '0'
+                }
+            }
+        ]
+        with mock.patch.object(RestClient, 'do_call', side_effect=ret):
+            ports = driver.list_ports(context)
+
+            self.assertDictEqual(ports[0], expected[0])
+            self.assertDictEqual(ports[1], expected[1])
+            self.assertDictEqual(ports[2], expected[2])
+            self.assertDictEqual(ports[3], expected[3])
+            self.assertDictEqual(ports[4], expected[4])
+            self.assertDictEqual(ports[5], expected[5])
+
+        with mock.patch.object(RestClient, 'get_all_ports',
+                               side_effect=exception.DelfinException):
+            with self.assertRaises(Exception) as exc:
+                driver.list_ports(context)
+            self.assertIn('An unknown exception occurred',
+                          str(exc.exception))
+
+        with mock.patch.object(RestClient, 'get_all_ports',
+                               side_effect=exception.InvalidResults):
+            with self.assertRaises(Exception) as exc:
+                driver.list_ports(context)
+            self.assertIn('The results are invalid',
+                          str(exc.exception))
+
+    def test_list_controllers(self):
+        driver = create_driver()
+        expected = [
+            {
+                'name': 'Controller-1',
+                'storage_id': '12345',
+                'native_controller_id': '0A',
+                'status': 'normal',
+                'location': 'Location1',
+                'soft_version': 'Ver123',
+                'cpu_info': 'Intel Xenon',
+                'memory_size': '100000',
+            },
+            {
+                'name': 'Controller-2',
+                'storage_id': '12345',
+                'native_controller_id': '0B',
+                'status': 'offline',
+                'location': 'Location2',
+                'soft_version': 'VerABC',
+                'cpu_info': 'ARM64',
+                'memory_size': '500000',
+            }
+
+        ]
+
+        ret = [
+            {
+                'data': [
```

Comment:
Can we have some more test data with all combination of Running status and helath status. that would give more coverage.

Reply:
Done